common/logger/src/sample.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implement periodic sampling for logs
+
+use std::{
+    sync::atomic::{AtomicU64, Ordering},
+    time::{Duration, SystemTime},
+};
+
+#[derive(Debug)]
+pub enum SampleRate {
+    Duration(Duration),
+    Frequency(u64),
+    Always,
+}
+
+pub struct Sampling {
+    rate: SampleRate,
+    state: AtomicU64,
+}
+
+impl Sampling {
+    pub const fn new(rate: SampleRate) -> Self {
+        Self {
+            rate,
+            state: AtomicU64::new(0),
+        }
+    }
+
+    pub fn sample(&self) -> bool {
+        match &self.rate {
+            SampleRate::Duration(rate) => Self::sample_duration(rate, &self.state),
+            SampleRate::Frequency(rate) => Self::sample_frequency(*rate, &self.state),
+            SampleRate::Always => true,
+        }
+    }
+
+    fn sample_frequency(rate: u64, count: &AtomicU64) -> bool {
+        let previous_count = count
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |count| {
+                let new_count = if count == 0 {
+                    rate.saturating_sub(1)
+                } else {
+                    count.saturating_sub(1)
+                };
+                Some(new_count)
+            })
+            .expect("Closure should always returns 'Some'. This is a Bug.");
+
+        previous_count == 0
+    }
+
+    fn sample_duration(rate: &Duration, last_sample: &AtomicU64) -> bool {
+        let rate = rate.as_secs();
+        // Seconds since Unix Epoch
+        let now = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("SystemTime before UNIX EPOCH!")
+            .as_secs();
```

Comment:
So, only by second granularity for sampling makes sense, let's enforce it in the input `Duration` and document it.  Also, I'd like to see if we make sampling `Duration` to restrict a duration of 0 seconds.